### ISSUE TITLE
[Bug Report] Not able to start scx programs with Kernel `6.8.1-cachyos-lto`

### ISSUE THREAD

**Issue by yqlbu** [sched-ext/scx#195]
## Summary

As the title suggests. It was working 3 days ago with Kernel `6.8.0-cachyos-lto` before I sync up the upstream changes.

## Addtional Info

Hardware:

```console
⯁ flake git:(master) ✗ ❯❯❯ neofetch
          ▗▄▄▄       ▗▄▄▄▄    ▄▄▄▖            kev@nixos-x1-carbon
          ▜███▙       ▜███▙  ▟███▛            -------------------
           ▜███▙       ▜███▙▟███▛             OS: NixOS 24.05.20240316.c75037b (Uakari) x86_64
            ▜███▙       ▜██████▛              Host: LENOVO 21KC0000CD
     ▟█████████████████▙ ▜████▛     ▟▙        Kernel: 6.8.1-cachyos
    ▟███████████████████▙ ▜███▙    ▟██▙       Uptime: 4 mins
           ▄▄▄▄▖           ▜███▙  ▟███▛       Packages: 1184 (nix-system), 1367 (nix-user), 6 (flatpak)
          ▟███▛             ▜██▛ ▟███▛        Shell: fish 3.7.0
         ▟███▛               ▜▛ ▟███▛         Resolution: 2880x1800
▟███████████▛                  ▟██████████▙   DE: Hyprland (Wayland)
▜██████████▛                  ▟███████████▛   WM: .Hyprland-wrapp
      ▟███▛ ▟▙               ▟███▛            Theme: Flat-Remix-GTK-Grey-Darkest [GTK2/3]
     ▟███▛ ▟██▙             ▟███▛             Icons: Papirus-Dark [GTK2/3]
    ▟███▛  ▜███▙           ▝▀▀▀▀              Terminal: alacritty
    ▜██▛    ▜███▙ ▜██████████████████▛        CPU: Intel Ultra 7 155H (22) @ 1.400GHz
     ▜▛     ▟████▙ ▜████████████████▛         GPU: Intel Arc Graphics]
           ▟██████▙       ▜███▙               Memory: 2921MiB / 31778MiB
          ▟███▛▜███▙       ▜███▙
         ▟███▛  ▜███▙       ▜███▙
         ▝▀▀▀    ▀▀▀▀▘       ▀▀▀▘
```

CPU: `Intel(R) Core(TM) Ultra 7 155H`

Platform: `NixOS 24.05.20240316.c75037b (Uakari) x86_64`

Kernel: `6.8.1-cachyos-lto`

Logs:

```console
 flake git:(master) ✗ ❯❯❯ sudo scx_rustland
thread 'main' panicked at src/main.rs:242:36:
Failed to build host topology: Failed to open or read file "/sys/devices/system/node/node0/cpu0/cache/index3/id"
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
✖ 101 flake git:(master) ✗ ❯❯❯ sudo scx_lavd
thread 'main' panicked at src/main.rs:115:36:
Failed to build host topology: Failed to open or read file "/sys/devices/system/node/node0/cpu0/cache/index3/id"
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by arighi** [comment]
Thanks for reporting this @yqlbu. Can you post the output of `lscpu -e`?

**Comment by ptr1337** [comment]
I have already talked about this with @PedroHLC and they did recently changes, that BTF was not included in their kernel, since they didnt package pahole as depend.

I think it should be already sorted out soon.

**Comment by arighi** [comment]
Hm.. weird that we're getting that topology error with BTF disabled (I was expecting a more generic error, like cannot start the scheduler or similar). But thanks for the update @ptr1337. 

@yqlbu can you double check that it's fixed now?

**Comment by ptr1337** [comment]
> Hm.. weird that we're getting that topology error with BTF disabled (I was expecting a more generic error, like cannot start the scheduler or similar). But thanks for the update @ptr1337.
> 
> @yqlbu can you double check that it's fixed now?

It could be also something different. Im not sure, which version Nix is currently using.
Lets wait for a report back from @PedroHLC and @yqlbu

**Comment by yqlbu** [comment]
> I have already talked about this with @PedroHLC and they did recently changes, that BTF was not included in their kernel, since they didnt package pahole as depend.
> 
> I think it should be already sorted out soon.
> 
> Should be fixed by: [chaotic-cx/nyx@10ab84a](https://github.com/chaotic-cx/nyx/commit/10ab84ad5dee3669f6152d92713e4d9ce8689241)

This bus has been fixed in https://github.com/chaotic-cx/nyx/pull/641 followed by the issue I raised https://github.com/chaotic-cx/nyx/issues/638

**Comment by yqlbu** [comment]
> Thanks for reporting this @yqlbu. Can you post the output of `lscpu -e`?

```console
⯁ ~ ❯❯❯ lscpu -e
CPU NODE SOCKET CORE L1d:L1i:L2:L3 ONLINE    MAXMHZ   MINMHZ       MHZ
  0    0      0    0 0:0:0:0          yes 4600.0000 400.0000 2310.3250
  1    0      0    0 0:0:0:0          yes 4600.0000 400.0000  879.1000
  2    0      0    1 4:4:1:0          yes 4600.0000 400.0000 3560.7019
  3    0      0    1 4:4:1:0          yes 4600.0000 400.0000 1955.7660
  4    0      0    2 8:8:2:0          yes 4600.0000 400.0000 4472.0791
  5    0      0    2 8:8:2:0          yes 4600.0000 400.0000 1492.7990
  6    0      0    3 12:12:3:0        yes 4600.0000 400.0000 3482.0840
  7    0      0    3 12:12:3:0        yes 4600.0000 400.0000 3396.1580
  8    0      0    4 16:16:4:0        yes 4700.0000 400.0000 2482.0171
  9    0      0    4 16:16:4:0        yes 4700.0000 400.0000 1915.4540
 10    0      0    5 20:20:5:0        yes 4700.0000 400.0000 4407.9160
 11    0      0    5 20:20:5:0        yes 4700.0000 400.0000 3873.0330
 12    0      0    6 24:24:6:0        yes 3500.0000 400.0000 1805.8820
 13    0      0    7 25:25:6:0        yes 3500.0000 400.0000 1424.6670
 14    0      0    8 26:26:6:0        yes 3500.0000 400.0000 1584.1460
 15    0      0    9 27:27:6:0        yes 3500.0000 400.0000 3225.4150
 16    0      0   10 28:28:7:0        yes 3500.0000 400.0000 2445.6470
 17    0      0   11 29:29:7:0        yes 3500.0000 400.0000  836.6460
 18    0      0   12 30:30:7:0        yes 3500.0000 400.0000 3101.6411
 19    0      0   13 31:31:7:0        yes 3500.0000 400.0000 2981.4141
⯁ ~ ❯❯❯
```

**Comment by yqlbu** [comment]
Surprisingly, the scx programs work fine with kernel `6.8.1-cachyos-lto` in my desktops but not laptop.

**Comment by ptr1337** [comment]
> Surprisingly, the scx programs work fine with kernel `6.8.1-cachyos-lto` in my desktops but not laptop.

Good to know. then it could be an topology issue with meteor lake?

**Comment by yqlbu** [comment]
> > Surprisingly, the scx programs work fine with kernel `6.8.1-cachyos-lto` in my desktops but not laptop.
> 
> Good to know. then it could be an topology issue with meteor lake?

Anything we can deal with this?

**Comment by yqlbu** [comment]
> It could be that the topology module doesn't handle properly the heterogeneous layout of this CPU. Can you also post the output of:
> 
> ```
> grep . /sys/devices/system/node/node*/cpu*/cache/index*/id
> ```

```console
⯁ ~ ❯❯❯ grep . /sys/devices/system/node/node*/cpu*/cache/index*/id
fish: No matches for wildcard '/sys/devices/system/node/node*/cpu*/cache/index*/id'. See `help wildcards-globbing`.
grep . /sys/devices/system/node/node*/cpu*/cache/index*/id
       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
```

**Comment by arighi** [comment]
Now I'm confused... you don't have CPU cache info in sysfs?

**Comment by yqlbu** [comment]
> Now I'm confused... you don't have CPU cache info in sysfs?

I have no idea...

**Comment by yqlbu** [comment]
@arighi Are there any critical changes made to the `scx` programs recently? It was working fine 3 days ago...

**Comment by arighi** [comment]
A quick and dirty workaround could be the following:
```
diff --git a/rust/scx_utils/src/topology.rs b/rust/scx_utils/src/topology.rs
index 4ea2ebd..14628dd 100644
--- a/rust/scx_utils/src/topology.rs
+++ b/rust/scx_utils/src/topology.rs
@@ -340,8 +340,9 @@ fn create_numa_nodes(online_mask: &Cpumask) -> Result<Vec<Node>> {
 
             // L3 cache ID
             let cache_path = cpu_path.join("cache");
+            // Use LLC 0 if we fail to detect cache hierarchy.
             let llc_id =
-                read_file_usize(&cache_path.join(format!("index{}", CACHE_LEVEL)).join("id"))?;
+                read_file_usize(&cache_path.join(format!("index{}", CACHE_LEVEL)).join("id")).unwrap_or(0);
 
             // Min and max frequencies. If the kernel is not compiled with
             // CONFIG_CPU_FREQ, just assume 0 for both frequencies.
```

This should be a valid fix for `scx_rustland` and `scx_lavd`, since they don't care about cache topology, but it could have an impact on `scx_rusty`, hence the workaround.

It'd be interesting to understand why you don't seem to have the cache topology info in sysfs to properly address this issue. I'll investigate more later today, in the meantime I think you can apply the patch above.

**Comment by arighi** [comment]
> @arighi Are there any critical changes made to the `scx` programs recently? It was working fine on my laptop 3 days ago...

Also, it doesn't seem to be an scx issue, more likely a kernel issue. We have some changes in the topology crate, but that was like 6 days ago, I don't see any relevant changes in the past 3 days.

**Comment by yqlbu** [comment]
> > @arighi Are there any critical changes made to the `scx` programs recently? It was working fine on my laptop 3 days ago...
> 
> Also, it doesn't seem to be an scx issue, more likely a kernel issue. We have some changes in the topology crate, but that was like 6 days ago, I don't see any relevant changes in the past 3 days.

cc @PedroHLC

**Comment by yqlbu** [comment]
> A quick and dirty workaround could be the following:
> 
> ```
> diff --git a/rust/scx_utils/src/topology.rs b/rust/scx_utils/src/topology.rs
> index 4ea2ebd..14628dd 100644
> --- a/rust/scx_utils/src/topology.rs
> +++ b/rust/scx_utils/src/topology.rs
> @@ -340,8 +340,9 @@ fn create_numa_nodes(online_mask: &Cpumask) -> Result<Vec<Node>> {
>  
>              // L3 cache ID
>              let cache_path = cpu_path.join("cache");
> +            // Use LLC 0 if we fail to detect cache hierarchy.
>              let llc_id =
> -                read_file_usize(&cache_path.join(format!("index{}", CACHE_LEVEL)).join("id"))?;
> +                read_file_usize(&cache_path.join(format!("index{}", CACHE_LEVEL)).join("id")).unwrap_or(0);
>  
>              // Min and max frequencies. If the kernel is not compiled with
>              // CONFIG_CPU_FREQ, just assume 0 for both frequencies.
> ```
> 
> This should be a valid fix for `scx_rustland` and `scx_lavd`, since they don't care about cache topology, but it could have an impact on `scx_rusty`, hence the workaround.
> 
> It'd be interesting to understand why you don't seem to have the cache topology info in sysfs to properly address this issue. I'll investigate more later today, in the meantime I think you can apply the patch above.

Appreciate the help for the workarounds. I will have a try on my end.

**Comment by PedroHLC** [comment]
> >  in the past 3 days.
> 
> cc @PedroHLC

In the past ~5 days we:
- pulled 6.8.1;
- bumped the sched-ext patch:
  - https://github.com/CachyOS/kernel-patches/commit/e133767ff871e019f850ef2c0a311b3f8f0bd290#diff-821dec123c4cf9eb32f7c6949ede17a387753c20186e90cb625d42259de44396
  - https://github.com/CachyOS/kernel-patches/commit/e9e9b8ab5f60f70261418a2110dd7f2104271603#diff-821dec123c4cf9eb32f7c6949ede17a387753c20186e90cb625d42259de44396
- Bumped scx from `0.1.7` to `0e53e7a00a9ade265dd8fc15d7c8a95fb83d011f` (adding ladv and rlfifo)
- And we accidentally removed `CONFIG_DEBUG_INFO_BTF` but returned it already.

You can see all the changes here: github.com/chaotic-cx/nyx/compare/73e16a06baf8a9bb273a131b10b142f620c2b610...main

And this is working fine in my Zen3 machine, will look at the Intel one later.

**Comment by yqlbu** [comment]
> > > in the past 3 days.
> > 
> > 
> > cc @PedroHLC
> 
> In the past ~5 days we:
> 
> * pulled 6.8.1;
> * bumped the sched-ext patch:
>   
>   * [CachyOS/kernel-patches@e133767#diff-821dec123c4cf9eb32f7c6949ede17a387753c20186e90cb625d42259de44396](https://github.com/CachyOS/kernel-patches/commit/e133767ff871e019f850ef2c0a311b3f8f0bd290#diff-821dec123c4cf9eb32f7c6949ede17a387753c20186e90cb625d42259de44396)
>   * [CachyOS/kernel-patches@e9e9b8a#diff-821dec123c4cf9eb32f7c6949ede17a387753c20186e90cb625d42259de44396](https://github.com/CachyOS/kernel-patches/commit/e9e9b8ab5f60f70261418a2110dd7f2104271603#diff-821dec123c4cf9eb32f7c6949ede17a387753c20186e90cb625d42259de44396)
> * Bumped scx from `0.1.7` to `0e53e7a00a9ade265dd8fc15d7c8a95fb83d011f` (adding lavd and rlfifo)
> * And we accidentally removed `CONFIG_DEBUG_INFO_BTF` but returned it already.
> 
> You can see all the changes here: [chaotic-cx/nyx@73e16a0...main](https://github.com/chaotic-cx/nyx/compare/73e16a06baf8a9bb273a131b10b142f620c2b610...main)
> 
> And this is working fine in my Zen3 machine, will look at the Intel one later.
> 
> **EDIT**: Tested `scx_rustland` and `scx_lavd` in my laptop (`i7-11800H`) and they ran without any issue.

Thanks for the notes. @arighi I think it is the issue with `meteor lake`. My desktop is running with `12700H` works fine with the HEAD commits in above repos (kernel-6.8.1-cachyos-lto + scx-unstable-0e53e7a00a9ade265dd8fc15d7c8a95fb83d011f).

**Comment by yqlbu** [comment]
Some additional info I can confirm was that:

![image](https://github.com/sched-ext/scx/assets/31861128/b7c4ef41-07af-4612-9799-47d88bcc1627)

`scx-0.1.7` runs well with kernel `6.8.1-cachyos-lto` from nyx repo.

**Comment by Byte-Lab** [comment]
> > In the past ~5 days we:
> > ```
> > * pulled 6.8.1;
> > 
> > * bumped the sched-ext patch:
> >   
> >   * [CachyOS/kernel-patches@e133767#diff-821dec123c4cf9eb32f7c6949ede17a387753c20186e90cb625d42259de44396](https://github.com/CachyOS/kernel-patches/commit/e133767ff871e019f850ef2c0a311b3f8f0bd290#diff-821dec123c4cf9eb32f7c6949ede17a387753c20186e90cb625d42259de44396)
> >   * [CachyOS/kernel-patches@e9e9b8a#diff-821dec123c4cf9eb32f7c6949ede17a387753c20186e90cb625d42259de44396](https://github.com/CachyOS/kernel-patches/commit/e9e9b8ab5f60f70261418a2110dd7f2104271603#diff-821dec123c4cf9eb32f7c6949ede17a387753c20186e90cb625d42259de44396)
> > 
> > * Bumped scx from `0.1.7` to `0e53e7a00a9ade265dd8fc15d7c8a95fb83d011f` (adding lavd and rlfifo)
> > ```
> 
> ...
> 
> Ok lots of changes happened in the past 3 days then. From 0.1.7 to [0e53e7a](https://github.com/sched-ext/scx/commit/0e53e7a00a9ade265dd8fc15d7c8a95fb83d011f) some schedulers (like `scx_rustland`) started to use a generic topology crate to detect the right CPU hierarchy (with cache information, CPUs, cores, etc.).
> 
> All these info are collected from sysfs, but in this particular case it seems that some sysfs files (` /sys/devices/system/node/node*/cpu*/cache/index*/id`) are expected to be found to properly determine the cache hierarchy, but they don't exist.
> 
> Honestly I don't know if this is expected for this particular CPU or if something is missing in the kernel or what.
> 
> I'll try to investigate a bit more later today, unless someone finds some useful info before me. :)

Thanks @arighi for looking into this. It's indeed weird that we don't have that CPU cache information on Meteor Lake. Let's apply your idea for a workaround to just fall back to cache 0. If some CPUs don't provide cache info, that seems like our only option.

**Comment by arighi** [comment]
ack, I'll send a PR later today

**Comment by Byte-Lab** [comment]
@yqlbu you should no longer run into this issue now that https://github.com/sched-ext/scx/pull/196 has landed

**Comment by yqlbu** [comment]
Hey @PedroHLC, could we sync the `scx` package to the HEAD sha in this repo please? I found difficulties compiling the binary on my end:

![image](https://github.com/sched-ext/scx/assets/31861128/ba65dc83-a7f0-49a9-b7a8-e328e8f49cec)

![image](https://github.com/sched-ext/scx/assets/31861128/c4311574-5476-4378-9cf6-d49b099567f7)

**Comment by htejun** [comment]
@yqlbu Any chance you can post the resulit of `hwloc-ls`?

**Comment by yqlbu** [comment]
> @yqlbu Any chance you can post the resulit of `hwloc-ls`?

Sure thing:

```console
[nix-shell:~]$ hwloc-ls
Machine (31GB total)
  Package L#0
    NUMANode L#0 (P#0 31GB)
    L3 L#0 (24MB)
      L2 L#0 (1280KB) + L1d L#0 (48KB) + L1i L#0 (32KB) + Core L#0
        PU L#0 (P#0)
        PU L#1 (P#1)
      L2 L#1 (1280KB) + L1d L#1 (48KB) + L1i L#1 (32KB) + Core L#1
        PU L#2 (P#2)
        PU L#3 (P#3)
      L2 L#2 (1280KB) + L1d L#2 (48KB) + L1i L#2 (32KB) + Core L#2
        PU L#4 (P#4)
        PU L#5 (P#5)
      L2 L#3 (1280KB) + L1d L#3 (48KB) + L1i L#3 (32KB) + Core L#3
        PU L#6 (P#6)
        PU L#7 (P#7)
      L2 L#4 (1280KB) + L1d L#4 (48KB) + L1i L#4 (32KB) + Core L#4
        PU L#8 (P#8)
        PU L#9 (P#9)
      L2 L#5 (1280KB) + L1d L#5 (48KB) + L1i L#5 (32KB) + Core L#5
        PU L#10 (P#10)
        PU L#11 (P#11)
      L2 L#6 (2048KB)
        L1d L#6 (32KB) + L1i L#6 (64KB) + Core L#6 + PU L#12 (P#12)
        L1d L#7 (32KB) + L1i L#7 (64KB) + Core L#7 + PU L#13 (P#13)
        L1d L#8 (32KB) + L1i L#8 (64KB) + Core L#8 + PU L#14 (P#14)
        L1d L#9 (32KB) + L1i L#9 (64KB) + Core L#9 + PU L#15 (P#15)
      L2 L#7 (2048KB)
        L1d L#10 (32KB) + L1i L#10 (64KB) + Core L#10 + PU L#16 (P#16)
        L1d L#11 (32KB) + L1i L#11 (64KB) + Core L#11 + PU L#17 (P#17)
        L1d L#12 (32KB) + L1i L#12 (64KB) + Core L#12 + PU L#18 (P#18)
        L1d L#13 (32KB) + L1i L#13 (64KB) + Core L#13 + PU L#19 (P#19)
  HostBridge
    PCIBridge
      PCIBridge
        PCIBridge
          PCI 03:00.0 (VGA)
    PCI 00:02.0 (VGA)
    PCIBridge
      PCI 05:00.0 (NVMExp)
        Block(Disk) "nvme0n1"
    PCIBridge
      PCIBridge
        PCIBridge
          PCI 3d:00.0 (Ethernet)
            Net "enp61s0"
    PCI 00:14.3 (Network)
      Net "wlo1"
    PCIBridge
      PCI 69:00.0 (Ethernet)
        Net "enp105s0"

[nix-shell:~]$
```

**Comment by htejun** [comment]
@yqlbu, can you also run the following command in bash and paste the output?
```bash
for i in $(ls -d /sys/devices/system/cpu/cpu*/cache/index* | sort -t / -k 6.4 -n); do printf "%16s: L%d %3s %s\n" ${i#/sys/devices/system/cpu/} $(cat $i/level) $(if [ -f $i/id ]; then cat $i/id; else echo "?"; fi) $(cat $i/shared_cpu_map); done
```

@Byte-Lab, the current topology implementation seems a bit broken. For example, the following is the `hwloc-ls` output on an m1 macbook:
```
[mac ~]$ hwloc-ls
Machine (15GB total)
  Package L#0
    NUMANode L#0 (P#0 15GB)
    L2 L#0 (4096KB)
      L1d L#0 (64KB) + L1i L#0 (128KB) + Core L#0 + PU L#0 (P#0)
      L1d L#1 (64KB) + L1i L#1 (128KB) + Core L#1 + PU L#1 (P#1)
    L2 L#1 (12MB)
      L1d L#2 (128KB) + L1i L#2 (192KB) + Core L#2 + PU L#2 (P#2)
      L1d L#3 (128KB) + L1i L#3 (192KB) + Core L#3 + PU L#3 (P#3)
      L1d L#4 (128KB) + L1i L#4 (192KB) + Core L#4 + PU L#4 (P#4)
    L2 L#2 (12MB)
      L1d L#5 (128KB) + L1i L#5 (192KB) + Core L#5 + PU L#5 (P#5)
      L1d L#6 (128KB) + L1i L#6 (192KB) + Core L#6 + PU L#6 (P#6)
      L1d L#7 (128KB) + L1i L#7 (192KB) + Core L#7 + PU L#7 (P#7)
  HostBridge
    PCIBridge
      PCI 01:00.0 (Network)
        Net "wlp1s0f0"
      PCI 01:00.1 (Network)
  Block(Disk) "nvme0n1"
  Block(Disk) "nvme0n2"
  Block(Disk) "nvme0n3"
```

If I run the above bash command:
```
[mac /sys/devices/system/cpu/cpu2/cache/index2]$ for i in $(ls -d /sys/devices/system/cpu/cpu*/cache/index* | sort -t / -k 6.4 -n); do printf "%16s: L%d %3s %s\n" ${i#/sys/devices/system/cpu/} $(cat $i/level) $(if [ -f $i/id ]; then cat $i/id; else echo "?"; fi) $(cat $i/shared_cpu_map); done
cpu0/cache/index0: L1   ? 01
cpu0/cache/index1: L1   ? 01
cpu0/cache/index2: L2   ? 03
cpu1/cache/index0: L1   ? 02
cpu1/cache/index1: L1   ? 02
cpu1/cache/index2: L2   ? 03
cpu2/cache/index0: L1   ? 04
cpu2/cache/index1: L1   ? 04
cpu2/cache/index2: L2   ? 1c
cpu3/cache/index0: L1   ? 08
cpu3/cache/index1: L1   ? 08
cpu3/cache/index2: L2   ? 1c
cpu4/cache/index0: L1   ? 10
cpu4/cache/index1: L1   ? 10
cpu4/cache/index2: L2   ? 1c
cpu5/cache/index0: L1   ? 20
cpu5/cache/index1: L1   ? 20
cpu5/cache/index2: L2   ? e0
cpu6/cache/index0: L1   ? 40
cpu6/cache/index1: L1   ? 40
cpu6/cache/index2: L2   ? e0
cpu7/cache/index0: L1   ? 80
cpu7/cache/index1: L1   ? 80
cpu7/cache/index2: L2   ? e0
```

- `indexN` may not necessarily indicate cache level N. That should come from `index*/level`.
- `id` may not exist. Looks like the sharing info should come from `shared_cpu_map`.
- wikipedia says that M1 chip has L3-ish system level cache shared with GPU but that doesn't seem reported anywhere. Doesn't really affect topology, so assuming everyone sharing if nothing is reported should work fine?

**Comment by yqlbu** [comment]
> @yqlbu, can you also run the following command in bash and paste the output?
> 
> ```shell
> for i in $(ls -d /sys/devices/system/cpu/cpu*/cache/index* | sort -t / -k 6.4 -n); do printf "%16s: L%d %3s %s\n" ${i#/sys/devices/system/cpu/} $(cat $i/level) $(if [ -f $i/id ]; then cat $i/id; else echo "?"; fi) $(cat $i/shared_cpu_map); done
> ```
> 
> @Byte-Lab, the current topology implementation seems a bit broken. For example, the following is the `hwloc-ls` output on an m1 macbook:
> 
> ```
> [mac ~]$ hwloc-ls
> Machine (15GB total)
>   Package L#0
>     NUMANode L#0 (P#0 15GB)
>     L2 L#0 (4096KB)
>       L1d L#0 (64KB) + L1i L#0 (128KB) + Core L#0 + PU L#0 (P#0)
>       L1d L#1 (64KB) + L1i L#1 (128KB) + Core L#1 + PU L#1 (P#1)
>     L2 L#1 (12MB)
>       L1d L#2 (128KB) + L1i L#2 (192KB) + Core L#2 + PU L#2 (P#2)
>       L1d L#3 (128KB) + L1i L#3 (192KB) + Core L#3 + PU L#3 (P#3)
>       L1d L#4 (128KB) + L1i L#4 (192KB) + Core L#4 + PU L#4 (P#4)
>     L2 L#2 (12MB)
>       L1d L#5 (128KB) + L1i L#5 (192KB) + Core L#5 + PU L#5 (P#5)
>       L1d L#6 (128KB) + L1i L#6 (192KB) + Core L#6 + PU L#6 (P#6)
>       L1d L#7 (128KB) + L1i L#7 (192KB) + Core L#7 + PU L#7 (P#7)
>   HostBridge
>     PCIBridge
>       PCI 01:00.0 (Network)
>         Net "wlp1s0f0"
>       PCI 01:00.1 (Network)
>   Block(Disk) "nvme0n1"
>   Block(Disk) "nvme0n2"
>   Block(Disk) "nvme0n3"
> ```
> 
> If I run the above bash command:
> 
> ```
> [mac /sys/devices/system/cpu/cpu2/cache/index2]$ for i in $(ls -d /sys/devices/system/cpu/cpu*/cache/index* | sort -t / -k 6.4 -n); do printf "%16s: L%d %3s %s\n" ${i#/sys/devices/system/cpu/} $(cat $i/level) $(if [ -f $i/id ]; then cat $i/id; else echo "?"; fi) $(cat $i/shared_cpu_map); done
> cpu0/cache/index0: L1   ? 01
> cpu0/cache/index1: L1   ? 01
> cpu0/cache/index2: L2   ? 03
> cpu1/cache/index0: L1   ? 02
> cpu1/cache/index1: L1   ? 02
> cpu1/cache/index2: L2   ? 03
> cpu2/cache/index0: L1   ? 04
> cpu2/cache/index1: L1   ? 04
> cpu2/cache/index2: L2   ? 1c
> cpu3/cache/index0: L1   ? 08
> cpu3/cache/index1: L1   ? 08
> cpu3/cache/index2: L2   ? 1c
> cpu4/cache/index0: L1   ? 10
> cpu4/cache/index1: L1   ? 10
> cpu4/cache/index2: L2   ? 1c
> cpu5/cache/index0: L1   ? 20
> cpu5/cache/index1: L1   ? 20
> cpu5/cache/index2: L2   ? e0
> cpu6/cache/index0: L1   ? 40
> cpu6/cache/index1: L1   ? 40
> cpu6/cache/index2: L2   ? e0
> cpu7/cache/index0: L1   ? 80
> cpu7/cache/index1: L1   ? 80
> cpu7/cache/index2: L2   ? e0
> ```
> 
> * `indexN` may not necessarily indicate cache level N. That should come from `index*/level`.
> * `id` may not exist. Looks like the sharing info should come from `shared_cpu_map`.
> * wikipedia says that M1 chip has L3-ish system level cache shared with GPU but that doesn't seem reported anywhere. Doesn't really affect topology, so assuming everyone sharing if nothing is reported should work fine?

```console
⯁ ~ ❯❯❯ chmod +x cpu.sh
⯁ ~ ❯❯❯ sh cpu.sh
cpu0/cache/index0: L1   0 00003
cpu0/cache/index1: L1   0 00003
cpu0/cache/index2: L2   0 00003
cpu0/cache/index3: L3   0 fffff
cpu1/cache/index0: L1   0 00003
cpu1/cache/index1: L1   0 00003
cpu1/cache/index2: L2   0 00003
cpu1/cache/index3: L3   0 fffff
cpu2/cache/index0: L1   4 0000c
cpu2/cache/index1: L1   4 0000c
cpu2/cache/index2: L2   1 0000c
cpu2/cache/index3: L3   0 fffff
cpu3/cache/index0: L1   4 0000c
cpu3/cache/index1: L1   4 0000c
cpu3/cache/index2: L2   1 0000c
cpu3/cache/index3: L3   0 fffff
cpu4/cache/index0: L1   8 00030
cpu4/cache/index1: L1   8 00030
cpu4/cache/index2: L2   2 00030
cpu4/cache/index3: L3   0 fffff
cpu5/cache/index0: L1   8 00030
cpu5/cache/index1: L1   8 00030
cpu5/cache/index2: L2   2 00030
cpu5/cache/index3: L3   0 fffff
cpu6/cache/index0: L1  12 000c0
cpu6/cache/index1: L1  12 000c0
cpu6/cache/index2: L2   3 000c0
cpu6/cache/index3: L3   0 fffff
cpu7/cache/index0: L1  12 000c0
cpu7/cache/index1: L1  12 000c0
cpu7/cache/index2: L2   3 000c0
cpu7/cache/index3: L3   0 fffff
cpu8/cache/index0: L1  16 00300
cpu8/cache/index1: L1  16 00300
cpu8/cache/index2: L2   4 00300
cpu8/cache/index3: L3   0 fffff
cpu9/cache/index0: L1  16 00300
cpu9/cache/index1: L1  16 00300
cpu9/cache/index2: L2   4 00300
cpu9/cache/index3: L3   0 fffff
cpu10/cache/index0: L1  20 00c00
cpu10/cache/index1: L1  20 00c00
cpu10/cache/index2: L2   5 00c00
cpu10/cache/index3: L3   0 fffff
cpu11/cache/index0: L1  20 00c00
cpu11/cache/index1: L1  20 00c00
cpu11/cache/index2: L2   5 00c00
cpu11/cache/index3: L3   0 fffff
cpu12/cache/index0: L1  24 01000
cpu12/cache/index1: L1  24 01000
cpu12/cache/index2: L2   6 0f000
cpu12/cache/index3: L3   0 fffff
cpu13/cache/index0: L1  25 02000
cpu13/cache/index1: L1  25 02000
cpu13/cache/index2: L2   6 0f000
cpu13/cache/index3: L3   0 fffff
cpu14/cache/index0: L1  26 04000
cpu14/cache/index1: L1  26 04000
cpu14/cache/index2: L2   6 0f000
cpu14/cache/index3: L3   0 fffff
cpu15/cache/index0: L1  27 08000
cpu15/cache/index1: L1  27 08000
cpu15/cache/index2: L2   6 0f000
cpu15/cache/index3: L3   0 fffff
cpu16/cache/index0: L1  28 10000
cpu16/cache/index1: L1  28 10000
cpu16/cache/index2: L2   7 f0000
cpu16/cache/index3: L3   0 fffff
cpu17/cache/index0: L1  29 20000
cpu17/cache/index1: L1  29 20000
cpu17/cache/index2: L2   7 f0000
cpu17/cache/index3: L3   0 fffff
cpu18/cache/index0: L1  30 40000
cpu18/cache/index1: L1  30 40000
cpu18/cache/index2: L2   7 f0000
cpu18/cache/index3: L3   0 fffff
cpu19/cache/index0: L1  31 80000
cpu19/cache/index1: L1  31 80000
cpu19/cache/index2: L2   7 f0000
cpu19/cache/index3: L3   0 fffff
```

**Comment by Byte-Lab** [comment]
@htejun ack, that all makes sense. The topology crate in its current form was in part taken from how we were detecting all of this in rusty before. I guess we never observed some of these issues before given not as many people were using it? I'll refactor to take into account what you're suggesting.

>wikipedia says that M1 chip has L3-ish system level cache shared with GPU but that doesn't seem reported anywhere. Doesn't really affect topology, so assuming everyone sharing if nothing is reported should work fine?

Yeah, seems like a reasonable implementation. There's not much we can do if it's not reported, and if we eventually find out that it's reported elsewhere we can always augment topology.rs to account for it.

**Comment by yqlbu** [comment]
As an update, I can confirm that https://github.com/sched-ext/scx/pull/196 does provide a valid workaround.

Results:

![image](https://github.com/sched-ext/scx/assets/31861128/87c39b18-679c-4b7b-b8eb-b369b49d6196)

**Comment by yqlbu** [comment]
I think we are good to close this issue? @Byte-Lab @arighi

**Comment by arighi** [comment]
Yes, this is fixed by 1bd990fb8747f39bdda9745b47b5a1718d2c3d83 for now.
